### PR TITLE
Add HTTP packet generator and fix bugs in old pkgen.

### PR DESCRIPTION
**Pull request by justinemarie**

### CODE REVIEW

File: bessctl/conf/perftest/pktgen_http.bess [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+assert(250 <= pkt_size <= 1522)
+assert(1 <= num_ports <= 16)
+assert(1 <= num_cores <= 4)
+
+# generate flows by varying IP dst 
+num_flows = int($SN_FLOWS!'1')
+assert(1 <= num_flows <= 10000)
+
+# Random ethernet address
+eth = scapy.Ether(src='02:1e:67:9f:4d:ae', dst='06:16:3e:1b:72:32')
```
Comment (melvinw) [inline bessctl/conf/perftest/pktgen_http.bess]: Could you make these configurable with env vars similar to SN_PORTS, SN_CORES, etc...? I think that will make using this in an automated way easier

File: bessctl/conf/perftest/pktgen_http.bess [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+# Increment this value
+ip = scapy.IP(src='192.168.0.1', dst='10.0.0.1')
```
Comment (melvinw) [inline bessctl/conf/perftest/pktgen_http.bess]: Same here

File: bessctl/conf/perftest/pktgen_http.bess [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+# HTTP flows, dst must be 80, src can be anything
+udp = scapy.TCP(sport=10001, dport=80)
```
Comment (melvinw) [inline bessctl/conf/perftest/pktgen_http.bess]: Here too :)

**Comment by melvinw** [comment]
Just a few comments about configurability. Looks good otherwise

File: bessctl/conf/perftest/pktgen_http.bess [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+    src = Source()
+    bess.attach_task(src.name, 0, wid=i)
+
+    rr = RoundRobin(gates=range(num_ports), mode="batch")
+    rw = Rewrite(templates=packet_templates_bytes)
+
+    src \
+    -> rw \
+    -> RandomUpdate(fields=[{'offset': 13, 'size': 2, 'min': 10002, 'max': 10002 + num_flows - 1}]) \
```
Comment (melvinw) [inline bessctl/conf/perftest/pktgen_http.bess]: `BESS_SRC_PORT + 1`?

Comment (justinemarie) [inline bessctl/conf/perftest/pktgen_http.bess]: That's into the IP field.

Comment (justinemarie) [inline bessctl/conf/perftest/pktgen_http.bess]: Oh I can fix that.

Comment (justinemarie) [inline bessctl/conf/perftest/pktgen_http.bess]: brb

File: bessctl/conf/perftest/pktgen_http.bess [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+# HTTP flows, dst must be 80, src can be anything
+src_port = int($BESS_SRC_PORT!'10001')
+tcp = scapy.TCP(sport=src_port, dport=80)
+
+packet_templates = []
+packet_templates_bytes = []
+
+for i in range(10):
+    host = 'http://www.' + chr(i + 97) * 3 + '.com'
+    payload = 'GET /pub/WWW/TheProject.html HTTP/1.1 \nHost: ' + host + '\n\n'
```
Comment (melvinw) [inline bessctl/conf/perftest/pktgen_http.bess]: "HTTP/1.1 \n" should be "HTTP/1.1\r\n"

Comment (melvinw) [inline bessctl/conf/perftest/pktgen_http.bess]: my parser is getting mad

File: bessctl/conf/perftest/pktgen_http.bess [hunk 6]
```diff
@@ -28,7 +28,7 @@ packet_templates_bytes = []
 
 for i in range(10):
     host = 'http://www.' + chr(i + 97) * 3 + '.com'
-    payload = 'GET /pub/WWW/TheProject.html HTTP/1.1 \nHost: ' + host + '\n\n'
+    payload = 'GET /pub/WWW/TheProject.html HTTP/1.1 \r\nHost: ' + host + '\r\n\n'
```
Comment (changlan) [inline bessctl/conf/perftest/pktgen_http.bess]: Shouldn't the end of header be '\r\n\r\n'?

**Comment by changlan** [comment]
Just realize that this HTTP generator won't work with @barath's TCP reconstruction code, because it does not do a valid three-way handshake (i.e. no SYN packet). Not sure whether scapy can generate do that.

**Comment by barath** [comment]
?  My TCP reconstruction code doesn't need a full handshake -- if you can
even send a fake SYN, that's all that's needed.  (You need a SYN because
the code needs to know the initial sequence number.)

**Comment by barath** [comment]
Stray "?"

**Comment by justinemarie** [comment]
No, Chang is right -- the packet generator isn't generating multi-packet
flows, every packet is a new "flow". I will make a patch...

**Comment by barath** [comment]
I agree.  All you'll need is to send a SYN before each if you have 1-packet
flows.

File: bessctl/conf/perftest/pktgen_http.bess [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+ip = scapy.IP(src=src_ip, dst=dst_ip)
+# HTTP flows, dst must be 80, src can be anything
+src_port = int($BESS_SRC_PORT!'10001')
+tcp = scapy.TCP(sport=src_port, dport=80)
+
+packet_templates = []
+packet_templates_bytes = []
+
+for i in range(10):
+    host = 'http://www.' + chr(i + 97) * 3 + '.com'
```
Comment (sangjinhan) [inline bessctl/conf/perftest/pktgen_http.bess]: You don't need `http://` for the HTTP Host header, do you?

Comment (melvinw) [inline bessctl/conf/perftest/pktgen_http.bess]: I think you're right. It looks like a `host` is defined as the part between "http://" and the path according to these.
https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.23
https://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html#sec3.2.2

File: bessctl/conf/perftest/pktgen_http.bess [hunk 8]
```diff
@@ -50,7 +50,7 @@ for i in range(num_cores):
     #rewrite the last two octets of the source IP.
     src \
     -> rw \
-    -> RandomUpdate(fields=[{'offset': 14, 'size': 2, 'min': 1, 'max': num_flows}]) \
+    -> RandomUpdate(fields=[{'offset': 30, 'size': 2, 'min': 1, 'max': num_flows}]) \
```
Comment (sangjinhan) [inline bessctl/conf/perftest/pktgen_http.bess]: The offset doesn't seem to match the comment. At offset 30 there is a destination IP address.

**Comment by justinemarie** [comment]
what do

**Pull request merged**